File: packages/test-cases/cases/bind/sanity/output/wasm-as/common/imported/TestImport_Object/serialization.ts
import {
  Read,
  ReadDecoder,
  Write,
  WriteSizer,
  WriteEncoder,
  Nullable,
  BigInt,
  BigNumber,
  JSON,
  Context
} from "@web3api/wasm-as";
import { TestImport_Object } from "./";
import * as Types from "../..";

export function serializeTestImport_Object(type: TestImport_Object): ArrayBuffer {
  const sizerContext: Context = new Context("Serializing (sizing)  imported object-type: TestImport_Object");
  const sizer = new WriteSizer(sizerContext);
  writeTestImport_Object(sizer, type);
  const buffer = new ArrayBuffer(sizer.length);
  const encoderContext: Context = new Context("Serializing (encoding) import object-type: TestImport_Object");
  const encoder = new WriteEncoder(buffer, sizer, encoderContext);
  writeTestImport_Object(encoder, type);
  return buffer;
}

export function writeTestImport_Object(writer: Write, type: TestImport_Object): void {
  writer.writeMapLength(12);
  writer.context().push("object", "Types.TestImport_AnotherObject", "writing property");
  writer.writeString("object");
  Types.TestImport_AnotherObject.write(writer, type.object);
  writer.context().pop();
  writer.context().push("optObject", "Types.TestImport_AnotherObject | null", "writing property");
  writer.writeString("optObject");
  if (type.optObject) {
    Types.TestImport_AnotherObject.write(writer, type.optObject as Types.TestImport_AnotherObject);
  } else {
    writer.writeNil();
  }
  writer.context().pop();
  writer.context().push("objectArray", "Array<Types.TestImport_AnotherObject>", "writing property");
  writer.writeString("objectArray");
  writer.writeArray(type.objectArray, (writer: Write, item: Types.TestImport_AnotherObject): void => {
    Types.TestImport_AnotherObject.write(writer, item);
  });
  writer.context().pop();
  writer.context().push("optObjectArray", "Array<Types.TestImport_AnotherObject | null> | null", "writing property");
  writer.writeString("optObjectArray");
  writer.writeNullableArray(type.optObjectArray, (writer: Write, item: Types.TestImport_AnotherObject | null): void => {
    if (item) {
      Types.TestImport_AnotherObject.write(writer, item as Types.TestImport_AnotherObject);
    } else {
      writer.writeNil();
    }
  });
  writer.context().pop();
  writer.context().push("en", "Types.TestImport_Enum", "writing property");
  writer.writeString("en");
  writer.writeInt32(type.en);
  writer.context().pop();
  writer.context().push("optEnum", "Nullable<Types.TestImport_Enum>", "writing property");
  writer.writeString("optEnum");
  writer.writeNullableInt32(type.optEnum);
  writer.context().pop();
  writer.context().push("enumArray", "Array<Types.TestImport_Enum>", "writing property");
  writer.writeString("enumArray");
  writer.writeArray(type.enumArray, (writer: Write, item: Types.TestImport_Enum): void => {
    writer.writeInt32(item);
  });
  writer.context().pop();
  writer.context().push("optEnumArray", "Array<Nullable<Types.TestImport_Enum>> | null", "writing property");
  writer.writeString("optEnumArray");
  writer.writeNullableArray(type.optEnumArray, (writer: Write, item: Nullable<Types.TestImport_Enum>): void => {
    writer.writeNullableInt32(item);
  });
  writer.context().pop();
  writer.context().push("union", "Types.TestImport_Union", "writing property");
  writer.writeString("union");
  Types.TestImport_Union.write(writer, type.union);
  writer.context().pop();
  writer.context().push("optUnion", "Types.TestImport_Union | null", "writing property");
  writer.writeString("optUnion");
  if (type.optUnion) {
    Types.TestImport_Union.write(writer, type.optUnion as Types.TestImport_Union);
  } else {
    writer.writeNil();
  }
  writer.context().pop();
  writer.context().push("unionArray", "Array<Types.TestImport_Union>", "writing property");
  writer.writeString("unionArray");
  writer.writeArray(type.unionArray, (writer: Write, item: Types.TestImport_Union): void => {
    Types.TestImport_Union.write(writer, item);
  });
  writer.context().pop();
  writer.context().push("optUnionArray", "Array<Types.TestImport_Union | null> | null", "writing property");
  writer.writeString("optUnionArray");
  writer.writeNullableArray(type.optUnionArray, (writer: Write, item: Types.TestImport_Union | null): void => {
    if (item) {
      Types.TestImport_Union.write(writer, item as Types.TestImport_Union);
    } else {
      writer.writeNil();
    }
  });
  writer.context().pop();
}

export function deserializeTestImport_Object(buffer: ArrayBuffer): TestImport_Object {
  const context: Context = new Context("Deserializing imported object-type TestImport_Object");
  const reader = new ReadDecoder(buffer, context);
  return readTestImport_Object(reader);
}

export function readTestImport_Object(reader: Read): TestImport_Object {
  let numFields = reader.readMapLength();

  let _object: Types.TestImport_AnotherObject | null = null;
  let _objectSet: bool = false;
  let _optObject: Types.TestImport_AnotherObject | null = null;
  let _objectArray: Array<Types.TestImport_AnotherObject> = [];
  let _objectArraySet: bool = false;
  let _optObjectArray: Array<Types.TestImport_AnotherObject | null> | null = null;
  let _en: Types.TestImport_Enum = 0;
  let _enSet: bool = false;
  let _optEnum: Nullable<Types.TestImport_Enum> = new Nullable<Types.TestImport_Enum>();
  let _enumArray: Array<Types.TestImport_Enum> = [];
  let _enumArraySet: bool = false;
  let _optEnumArray: Array<Nullable<Types.TestImport_Enum>> | null = null;
  let _union: Types.TestImport_Union | null = null;
  let _unionSet: bool = false;
  let _optUnion: Types.TestImport_Union | null = null;
  let _unionArray: Array<Types.TestImport_Union> = [];
  let _unionArraySet: bool = false;
  let _optUnionArray: Array<Types.TestImport_Union | null> | null = null;

  while (numFields > 0) {
    numFields--;
    const field = reader.readString();

    reader.context().push(field, "unknown", "searching for property type");
    if (field == "object") {
      reader.context().push(field, "Types.TestImport_AnotherObject", "type found, reading property");
      const object = Types.TestImport_AnotherObject.read(reader);
      _object = object;
      _objectSet = true;
      reader.context().pop();
    }
    else if (field == "optObject") {
      reader.context().push(field, "Types.TestImport_AnotherObject | null", "type found, reading property");
      let object: Types.TestImport_AnotherObject | null = null;
      if (!reader.isNextNil()) {
        object = Types.TestImport_AnotherObject.read(reader);
      }
      _optObject = object;
      reader.context().pop();
    }
    else if (field == "objectArray") {
      reader.context().push(field, "Array<Types.TestImport_AnotherObject>", "type found, reading property");
      _objectArray = reader.readArray((reader: Read): Types.TestImport_AnotherObject => {
        const object = Types.TestImport_AnotherObject.read(reader);
        return object;
      });
      _objectArraySet = true;
      reader.context().pop();
    }
    else if (field == "optObjectArray") {
      reader.context().push(field, "Array<Types.TestImport_AnotherObject | null> | null", "type found, reading property");
      _optObjectArray = reader.readNullableArray((reader: Read): Types.TestImport_AnotherObject | null => {
        let object: Types.TestImport_AnotherObject | null = null;
        if (!reader.isNextNil()) {
          object = Types.TestImport_AnotherObject.read(reader);
        }
        return object;
      });
      reader.context().pop();
    }
    else if (field == "en") {
      reader.context().push(field, "Types.TestImport_Enum", "type found, reading property");
      let value: Types.TestImport_Enum;
      if (reader.isNextString()) {
        value = Types.getTestImport_EnumValue(reader.readString());
      } else {
        value = reader.readInt32();
        Types.sanitizeTestImport_EnumValue(value);
      }
      _en = value;
      _enSet = true;
      reader.context().pop();
    }
    else if (field == "optEnum") {
      reader.context().push(field, "Nullable<Types.TestImport_Enum>", "type found, reading property");
      let value: Nullable<Types.TestImport_Enum>;
      if (!reader.isNextNil()) {
        if (reader.isNextString()) {
          value = Nullable.fromValue(
            Types.getTestImport_EnumValue(reader.readString())
          );
        } else {
          value = Nullable.fromValue(
            reader.readInt32()
          );
          Types.sanitizeTestImport_EnumValue(value.value);
        }
      } else {
        value = Nullable.fromNull<Types.TestImport_Enum>();
      }
      _optEnum = value;
      reader.context().pop();
    }
    else if (field == "enumArray") {
      reader.context().push(field, "Array<Types.TestImport_Enum>", "type found, reading property");
      _enumArray = reader.readArray((reader: Read): Types.TestImport_Enum => {
        let value: Types.TestImport_Enum;
        if (reader.isNextString()) {
          value = Types.getTestImport_EnumValue(reader.readString());
        } else {
          value = reader.readInt32();
          Types.sanitizeTestImport_EnumValue(value);
        }
        return value;
      });
      _enumArraySet = true;
      reader.context().pop();
    }
    else if (field == "optEnumArray") {
      reader.context().push(field, "Array<Nullable<Types.TestImport_Enum>> | null", "type found, reading property");
      _optEnumArray = reader.readNullableArray((reader: Read): Nullable<Types.TestImport_Enum> => {
        let value: Nullable<Types.TestImport_Enum>;
        if (!reader.isNextNil()) {
          if (reader.isNextString()) {
            value = Nullable.fromValue(
              Types.getTestImport_EnumValue(reader.readString())
            );
          } else {
            value = Nullable.fromValue(
              reader.readInt32()
            );
            Types.sanitizeTestImport_EnumValue(value.value);
          }
        } else {
          value = Nullable.fromNull<Types.TestImport_Enum>();
        }
        return value;
      });
      reader.context().pop();
    }
    else if (field == "union") {
      reader.context().push(field, "Types.TestImport_Union", "type found, reading property");
      const union = Types.TestImport_Union.read(reader);
      _union = union;
      _unionSet = true;
      reader.context().pop();
    }
    else if (field == "optUnion") {
      reader.context().push(field, "Types.TestImport_Union | null", "type found, reading property");
      let union: Types.TestImport_Union | null = null;
      if (!reader.isNextNil()) {
        union = Types.TestImport_Union.read(reader);
      }
      _optUnion = union;
      reader.context().pop();
    }
    else if (field == "unionArray") {
      reader.context().push(field, "Array<Types.TestImport_Union>", "type found, reading property");
      _unionArray = reader.readArray((reader: Read): Types.TestImport_Union => {
        const union = Types.TestImport_Union.read(reader);
        return union;
      });
      _unionArraySet = true;
      reader.context().pop();
    }
    else if (field == "optUnionArray") {
      reader.context().push(field, "Array<Types.TestImport_Union | null> | null", "type found, reading property");
      _optUnionArray = reader.readNullableArray((reader: Read): Types.TestImport_Union | null => {
        let union: Types.TestImport_Union | null = null;
        if (!reader.isNextNil()) {
          union = Types.TestImport_Union.read(reader);
        }
        return union;
      });
      reader.context().pop();
    }
    reader.context().pop();
  }

  if (!_object || !_objectSet) {
    throw new Error(reader.context().printWithContext("Missing required property: 'object: TestImport_AnotherObject'"));
  }
  if (!_objectArraySet) {
    throw new Error(reader.context().printWithContext("Missing required property: 'objectArray: [TestImport_AnotherObject]'"));
  }
  if (!_enSet) {
    throw new Error(reader.context().printWithContext("Missing required property: 'en: TestImport_Enum'"));
  }
  if (!_enumArraySet) {
    throw new Error(reader.context().printWithContext("Missing required property: 'enumArray: [TestImport_Enum]'"));
  }
  if (!_union || !_unionSet) {
    throw new Error(reader.context().printWithContext("Missing required property: 'union: TestImport_Union'"));
  }
  if (!_unionArraySet) {
    throw new Error(reader.context().printWithContext("Missing required property: 'unionArray: [TestImport_Union]'"));
  }

  return {
    object: _object,
    optObject: _optObject,
    objectArray: _objectArray,
    optObjectArray: _optObjectArray,
    en: _en,
    optEnum: _optEnum,
    enumArray: _enumArray,
    optEnumArray: _optEnumArray,
    union: _union,
    optUnion: _optUnion,
    unionArray: _unionArray,
    optUnionArray: _optUnionArray
  };
}

File: packages/test-cases/cases/bind/sanity/output/wasm-rs/module/serialization.rs
use serde::{Serialize, Deserialize};
use std::convert::TryFrom;
use polywrap_wasm_rs::{
    BigInt,
    BigNumber,
    Map,
    Context,
    DecodeError,
    EncodeError,
    Read,
    ReadDecoder,
    Write,
    WriteEncoder,
    JSON,
};

use crate::{
    CustomEnum,
    get_custom_enum_value,
    sanitize_custom_enum_value
};
use crate::AnotherType;
use crate::Else;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ArgsModuleMethod {
    pub str: String,
    pub opt_str: Option<String>,
    pub en: CustomEnum,
    pub opt_enum: Option<CustomEnum>,
    pub enum_array: Vec<CustomEnum>,
    pub opt_enum_array: Option<Vec<Option<CustomEnum>>>,
    pub map: Map<String, i32>,
    pub map_of_arr: Map<String, Vec<i32>>,
    pub map_of_map: Map<String, Map<String, i32>>,
    pub map_of_obj: Map<String, AnotherType>,
    pub map_of_arr_of_obj: Map<String, Vec<AnotherType>>,
}

pub fn deserialize_module_method_args(args: &[u8]) -> Result<ArgsModuleMethod, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing module-type: module_method".to_string();

    let mut reader = ReadDecoder::new(args, context);
    let mut num_of_fields = reader.read_map_length()?;

    let mut _str: String = String::new();
    let mut _str_set = false;
    let mut _opt_str: Option<String> = None;
    let mut _en: CustomEnum = CustomEnum::_MAX_;
    let mut _en_set = false;
    let mut _opt_enum: Option<CustomEnum> = None;
    let mut _enum_array: Vec<CustomEnum> = vec![];
    let mut _enum_array_set = false;
    let mut _opt_enum_array: Option<Vec<Option<CustomEnum>>> = None;
    let mut _map: Map<String, i32> = Map::<String, i32>::new();
    let mut _map_set = false;
    let mut _map_of_arr: Map<String, Vec<i32>> = Map::<String, Vec<i32>>::new();
    let mut _map_of_arr_set = false;
    let mut _map_of_map: Map<String, Map<String, i32>> = Map::<String, Map<String, i32>>::new();
    let mut _map_of_map_set = false;
    let mut _map_of_obj: Map<String, AnotherType> = Map::<String, AnotherType>::new();
    let mut _map_of_obj_set = false;
    let mut _map_of_arr_of_obj: Map<String, Vec<AnotherType>> = Map::<String, Vec<AnotherType>>::new();
    let mut _map_of_arr_of_obj_set = false;

    while num_of_fields > 0 {
        num_of_fields -= 1;
        let field = reader.read_string()?;

        match field.as_str() {
            "str" => {
                reader.context().push(&field, "String", "type found, reading argument");
                _str = reader.read_string()?;
                _str_set = true;
                reader.context().pop();
            }
            "optStr" => {
                reader.context().push(&field, "Option<String>", "type found, reading argument");
                _opt_str = reader.read_optional_string()?;
                reader.context().pop();
            }
            "en" => {
                reader.context().push(&field, "CustomEnum", "type found, reading argument");
                let mut value: CustomEnum = CustomEnum::_MAX_;
                if reader.is_next_string()? {
                    value = get_custom_enum_value(&reader.read_string()?)?;
                } else {
                    value = CustomEnum::try_from(reader.read_i32()?)?;
                    sanitize_custom_enum_value(value as i32)?;
                }
                _en = value;
                _en_set = true;
                reader.context().pop();
            }
            "optEnum" => {
                reader.context().push(&field, "Option<CustomEnum>", "type found, reading argument");
                let mut value: Option<CustomEnum> = None;
                if !reader.is_next_nil()? {
                    if reader.is_next_string()? {
                        value = Some(get_custom_enum_value(&reader.read_string()?)?);
                    } else {
                        value = Some(CustomEnum::try_from(reader.read_i32()?)?);
                        sanitize_custom_enum_value(value.unwrap() as i32)?;
                    }
                } else {
                    value = None;
                }
                _opt_enum = value;
                reader.context().pop();
            }
            "enumArray" => {
                reader.context().push(&field, "Vec<CustomEnum>", "type found, reading argument");
                _enum_array = reader.read_array(|reader| {
                    let mut value: CustomEnum = CustomEnum::_MAX_;
                    if reader.is_next_string()? {
                        value = get_custom_enum_value(&reader.read_string()?)?;
                    } else {
                        value = CustomEnum::try_from(reader.read_i32()?)?;
                        sanitize_custom_enum_value(value as i32)?;
                    }
                    Ok(value)
                })?;
                _enum_array_set = true;
                reader.context().pop();
            }
            "optEnumArray" => {
                reader.context().push(&field, "Option<Vec<Option<CustomEnum>>>", "type found, reading argument");
                _opt_enum_array = reader.read_optional_array(|reader| {
                    let mut value: Option<CustomEnum> = None;
                    if !reader.is_next_nil()? {
                        if reader.is_next_string()? {
                            value = Some(get_custom_enum_value(&reader.read_string()?)?);
                        } else {
                            value = Some(CustomEnum::try_from(reader.read_i32()?)?);
                            sanitize_custom_enum_value(value.unwrap() as i32)?;
                        }
                    } else {
                        value = None;
                    }
                    Ok(value)
                })?;
                reader.context().pop();
            }
            "map" => {
                reader.context().push(&field, "Map<String, i32>", "type found, reading argument");
                _map = reader.read_ext_generic_map(|reader| {
                    reader.read_string()
                }, |reader| {
                    reader.read_i32()
                })?;
                _map_set = true;
                reader.context().pop();
            }
            "mapOfArr" => {
                reader.context().push(&field, "Map<String, Vec<i32>>", "type found, reading argument");
                _map_of_arr = reader.read_ext_generic_map(|reader| {
                    reader.read_string()
                }, |reader| {
                    reader.read_array(|reader| {
                        reader.read_i32()
                    })
                })?;
                _map_of_arr_set = true;
                reader.context().pop();
            }
            "mapOfMap" => {
                reader.context().push(&field, "Map<String, Map<String, i32>>", "type found, reading argument");
                _map_of_map = reader.read_ext_generic_map(|reader| {
                    reader.read_string()
                }, |reader| {
                    reader.read_ext_generic_map(|reader| {
                        reader.read_string()
                    }, |reader| {
                        reader.read_i32()
                    })
                })?;
                _map_of_map_set = true;
                reader.context().pop();
            }
            "mapOfObj" => {
                reader.context().push(&field, "Map<String, AnotherType>", "type found, reading argument");
                _map_of_obj = reader.read_ext_generic_map(|reader| {
                    reader.read_string()
                }, |reader| {
                    let object = AnotherType::read(reader)?;
                    Ok(object)
                })?;
                _map_of_obj_set = true;
                reader.context().pop();
            }
            "mapOfArrOfObj" => {
                reader.context().push(&field, "Map<String, Vec<AnotherType>>", "type found, reading argument");
                _map_of_arr_of_obj = reader.read_ext_generic_map(|reader| {
                    reader.read_string()
                }, |reader| {
                    reader.read_array(|reader| {
                        let object = AnotherType::read(reader)?;
                        Ok(object)
                    })
                })?;
                _map_of_arr_of_obj_set = true;
                reader.context().pop();
            }
            err => return Err(DecodeError::UnknownFieldName(err.to_string())),
        }
    }
    if !_str_set {
        return Err(DecodeError::MissingField("str: String.".to_string()));
    }
    if !_en_set {
        return Err(DecodeError::MissingField("en: CustomEnum.".to_string()));
    }
    if !_enum_array_set {
        return Err(DecodeError::MissingField("enumArray: [CustomEnum].".to_string()));
    }
    if !_map_set {
        return Err(DecodeError::MissingField("map: Map<String, Int>.".to_string()));
    }
    if !_map_of_arr_set {
        return Err(DecodeError::MissingField("mapOfArr: Map<String, [Int]>.".to_string()));
    }
    if !_map_of_map_set {
        return Err(DecodeError::MissingField("mapOfMap: Map<String, Map<String, Int>>.".to_string()));
    }
    if !_map_of_obj_set {
        return Err(DecodeError::MissingField("mapOfObj: Map<String, AnotherType>.".to_string()));
    }
    if !_map_of_arr_of_obj_set {
        return Err(DecodeError::MissingField("mapOfArrOfObj: Map<String, [AnotherType]>.".to_string()));
    }

    Ok(ArgsModuleMethod {
        str: _str,
        opt_str: _opt_str,
        en: _en,
        opt_enum: _opt_enum,
        enum_array: _enum_array,
        opt_enum_array: _opt_enum_array,
        map: _map,
        map_of_arr: _map_of_arr,
        map_of_map: _map_of_map,
        map_of_obj: _map_of_obj,
        map_of_arr_of_obj: _map_of_arr_of_obj,
    })
}

pub fn serialize_module_method_result(result: &i32) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) module-type: module_method".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_module_method_result(result, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_module_method_result<W: Write>(result: &i32, writer: &mut W) -> Result<(), EncodeError> {
    writer.context().push("moduleMethod", "i32", "writing result");
    writer.write_i32(result)?;
    writer.context().pop();
    Ok(())
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ArgsObjectMethod {
    pub object: AnotherType,
    pub opt_object: Option<AnotherType>,
    pub object_array: Vec<AnotherType>,
    pub opt_object_array: Option<Vec<Option<AnotherType>>>,
}

pub fn deserialize_object_method_args(args: &[u8]) -> Result<ArgsObjectMethod, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing module-type: object_method".to_string();

    let mut reader = ReadDecoder::new(args, context);
    let mut num_of_fields = reader.read_map_length()?;

    let mut _object: AnotherType = AnotherType::new();
    let mut _object_set = false;
    let mut _opt_object: Option<AnotherType> = None;
    let mut _object_array: Vec<AnotherType> = vec![];
    let mut _object_array_set = false;
    let mut _opt_object_array: Option<Vec<Option<AnotherType>>> = None;

    while num_of_fields > 0 {
        num_of_fields -= 1;
        let field = reader.read_string()?;

        match field.as_str() {
            "object" => {
                reader.context().push(&field, "AnotherType", "type found, reading argument");
                let object = AnotherType::read(&mut reader)?;
                _object = object;
                _object_set = true;
                reader.context().pop();
            }
            "optObject" => {
                reader.context().push(&field, "Option<AnotherType>", "type found, reading argument");
                let mut object: Option<AnotherType> = None;
                if !reader.is_next_nil()? {
                    object = Some(AnotherType::read(&mut reader)?);
                } else {
                    object = None;
                }
                _opt_object = object;
                reader.context().pop();
            }
            "objectArray" => {
                reader.context().push(&field, "Vec<AnotherType>", "type found, reading argument");
                _object_array = reader.read_array(|reader| {
                    let object = AnotherType::read(reader)?;
                    Ok(object)
                })?;
                _object_array_set = true;
                reader.context().pop();
            }
            "optObjectArray" => {
                reader.context().push(&field, "Option<Vec<Option<AnotherType>>>", "type found, reading argument");
                _opt_object_array = reader.read_optional_array(|reader| {
                    let mut object: Option<AnotherType> = None;
                    if !reader.is_next_nil()? {
                        object = Some(AnotherType::read(reader)?);
                    } else {
                        object = None;
                    }
                    Ok(object)
                })?;
                reader.context().pop();
            }
            err => return Err(DecodeError::UnknownFieldName(err.to_string())),
        }
    }
    if !_object_set {
        return Err(DecodeError::MissingField("object: AnotherType.".to_string()));
    }
    if !_object_array_set {
        return Err(DecodeError::MissingField("objectArray: [AnotherType].".to_string()));
    }

    Ok(ArgsObjectMethod {
        object: _object,
        opt_object: _opt_object,
        object_array: _object_array,
        opt_object_array: _opt_object_array,
    })
}

pub fn serialize_object_method_result(result: &Option<AnotherType>) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) module-type: object_method".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_object_method_result(result, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_object_method_result<W: Write>(result: &Option<AnotherType>, writer: &mut W) -> Result<(), EncodeError> {
    writer.context().push("objectMethod", "Option<AnotherType>", "writing result");
    if result.is_some() {
        AnotherType::write(result.as_ref().unwrap(), writer)?;
    } else {
        writer.write_nil()?;
    }
    writer.context().pop();
    Ok(())
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ArgsOptionalEnvMethod {
    pub object: AnotherType,
    pub opt_object: Option<AnotherType>,
    pub object_array: Vec<AnotherType>,
    pub opt_object_array: Option<Vec<Option<AnotherType>>>,
}

pub fn deserialize_optional_env_method_args(args: &[u8]) -> Result<ArgsOptionalEnvMethod, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing module-type: optional_env_method".to_string();

    let mut reader = ReadDecoder::new(args, context);
    let mut num_of_fields = reader.read_map_length()?;

    let mut _object: AnotherType = AnotherType::new();
    let mut _object_set = false;
    let mut _opt_object: Option<AnotherType> = None;
    let mut _object_array: Vec<AnotherType> = vec![];
    let mut _object_array_set = false;
    let mut _opt_object_array: Option<Vec<Option<AnotherType>>> = None;

    while num_of_fields > 0 {
        num_of_fields -= 1;
        let field = reader.read_string()?;

        match field.as_str() {
            "object" => {
                reader.context().push(&field, "AnotherType", "type found, reading argument");
                let object = AnotherType::read(&mut reader)?;
                _object = object;
                _object_set = true;
                reader.context().pop();
            }
            "optObject" => {
                reader.context().push(&field, "Option<AnotherType>", "type found, reading argument");
                let mut object: Option<AnotherType> = None;
                if !reader.is_next_nil()? {
                    object = Some(AnotherType::read(&mut reader)?);
                } else {
                    object = None;
                }
                _opt_object = object;
                reader.context().pop();
            }
            "objectArray" => {
                reader.context().push(&field, "Vec<AnotherType>", "type found, reading argument");
                _object_array = reader.read_array(|reader| {
                    let object = AnotherType::read(reader)?;
                    Ok(object)
                })?;
                _object_array_set = true;
                reader.context().pop();
            }
            "optObjectArray" => {
                reader.context().push(&field, "Option<Vec<Option<AnotherType>>>", "type found, reading argument");
                _opt_object_array = reader.read_optional_array(|reader| {
                    let mut object: Option<AnotherType> = None;
                    if !reader.is_next_nil()? {
                        object = Some(AnotherType::read(reader)?);
                    } else {
                        object = None;
                    }
                    Ok(object)
                })?;
                reader.context().pop();
            }
            err => return Err(DecodeError::UnknownFieldName(err.to_string())),
        }
    }
    if !_object_set {
        return Err(DecodeError::MissingField("object: AnotherType.".to_string()));
    }
    if !_object_array_set {
        return Err(DecodeError::MissingField("objectArray: [AnotherType].".to_string()));
    }

    Ok(ArgsOptionalEnvMethod {
        object: _object,
        opt_object: _opt_object,
        object_array: _object_array,
        opt_object_array: _opt_object_array,
    })
}

pub fn serialize_optional_env_method_result(result: &Option<AnotherType>) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) module-type: optional_env_method".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_optional_env_method_result(result, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_optional_env_method_result<W: Write>(result: &Option<AnotherType>, writer: &mut W) -> Result<(), EncodeError> {
    writer.context().push("optionalEnvMethod", "Option<AnotherType>", "writing result");
    if result.is_some() {
        AnotherType::write(result.as_ref().unwrap(), writer)?;
    } else {
        writer.write_nil()?;
    }
    writer.context().pop();
    Ok(())
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ArgsIf {
    pub _if: Else,
}

pub fn deserialize_if_args(args: &[u8]) -> Result<ArgsIf, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing module-type: if".to_string();

    let mut reader = ReadDecoder::new(args, context);
    let mut num_of_fields = reader.read_map_length()?;

    let mut _if: Else = Else::new();
    let mut _if_set = false;

    while num_of_fields > 0 {
        num_of_fields -= 1;
        let field = reader.read_string()?;

        match field.as_str() {
            "if" => {
                reader.context().push(&field, "Else", "type found, reading argument");
                let object = Else::read(&mut reader)?;
                _if = object;
                _if_set = true;
                reader.context().pop();
            }
            err => return Err(DecodeError::UnknownFieldName(err.to_string())),
        }
    }
    if !_if_set {
        return Err(DecodeError::MissingField("if: else.".to_string()));
    }

    Ok(ArgsIf {
        _if: _if,
    })
}

pub fn serialize_if_result(result: &Else) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) module-type: if".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_if_result(result, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_if_result<W: Write>(result: &Else, writer: &mut W) -> Result<(), EncodeError> {
    writer.context().push("if", "Else", "writing result");
    Else::write(&result, writer)?;
    writer.context().pop();
    Ok(())
}
